### PR TITLE
General refactoring and automatic config accumulators

### PR DESCRIPTION
**Pull request by bmhughes**
# Description

- Set resource configuration Hash programmatically
- Automatic config Hash path creation
- Automatic resource property fetching
- Automatic config template resource generation
- Add service resource
- Add LDAP log filters to the main config file

## Issues Resolved

- #380 
- #388
- #392 
- #397 
- #400 
- #401 

## Check List

- [x] All tests pass. See TESTING.md for details.
- [x] New functionality includes testing.
- [x] New functionality has been documented in the README if applicable.

### CODE REVIEW

**Comment by kitchen-porter** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big Pull Request.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a Table Flip.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Review by damacus — DISMISSED** [review]

**Comment by damacus** [comment]
@bmhughes Looks like we've got the conflicts here

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,20 +4,12 @@ This file is used to list changes made in each version of grafana.
...
-
-All Grafana API resources will be removed in a future version of this cookbook. Other tools e.g. Terraform are better
-suited for interacting with the API.
-
-- `grafana_dashboard`, `grafana_dashboard_template`, `grafana_dashboard_backup`
-  - These use the old dashboard name API which was removed in the latest 8.0 release of Grafana
-- `grafana_datasource`, `grafana_datasource_backup`
-- `grafana_organization`
-- `grafana_user`
+- Set resource configuration Hash programmatically
```
Comment (jamie-zaharopoulos-fh) [inline CHANGELOG.md]: Was it intentional to remove the changeling for `## 9.7.0 - *2021-07-23*` ?

**Comment by bmhughes** [comment]
> 
> 
> @bmhughes Looks like we've got the conflicts here

Consider how many merge conflicts I ran into rebasing this I'm surprised it was only those two 😂

**Comment by bmhughes** [comment]
Converting this to a draft while I add `load_current_value` support to it

File: libraries/config.rb [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        false
+      end
+
+      def init_config_template(load_existing = false)
+        return false if config_template_exist?
+
+        Chef::Log.debug("init_config_template: Creating config template resource for #{new_resource.config_file}")
+
+        with_run_context(:root) do
+          declare_resource(:chef_gem, 'deepsort') { compile_time true } unless gem_installed?('deepsort')
```
Comment (damacus) [inline libraries/config.rb:127]: Do we still need compile time now we're using unified_mode?

File: DEVELOPING.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
+# Developing
+
+The refactor of the cookbooks in v10 makes some major changes to the underlying method of operation of the cookbook to improve operation and maintainabilty.
```
Comment (damacus) [inline DEVELOPING.md]:
```suggestion
The refactor for version 10 of this cookbook makes some major changes to the underlying method of operation of the cookbook to improve operation and maintainability.
```

File: DEVELOPING.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+use 'partial/_config_file'
+```
+
+At the top of the resource declaration above the first property.
+
+See a pre-existing resource for an example of this pattern.
+
+### Overiding the automatic path
+
+If the name of the resource cannot match the name of the configuration section or a nested configuration section is in use such as `section.subsection` then the automatic configuration path can be overriden by the addition of the `:resource_config_path_override` method. Due to needed to override the path in both the action class and the outer resource class this method must be added to both sections of the resource declaration.
```
Comment (damacus) [inline DEVELOPING.md]:
```suggestion
If the name of the resource does not match the name of the configuration section or a nested configuration section is in use such as `section.subsection` then the automatic configuration path can be overriden by the use of the `:resource_config_path_override` method. Due to needing to override the path in both the action class and the outer resource class this method must be added to both sections of the resource declaration.
```

File: DEVELOPING.md [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+Below the last property add the following with the array contents set to the config Hash path set as you would path to [Hash#dig](https://ruby-doc.org/core-3.0.2/Hash.html#method-i-dig) to create the (nested if required) configuration path.
```
Comment (damacus) [inline DEVELOPING.md]: I'm not sure this sentence makes sense right now. Are you able to reword it please?

File: DEVELOPING.md [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+end
+
+action_class do
+  def resource_config_path_override
+    %w(external_image_storage.s3)
+  end
+end
+```
+
+### Excluding properties from the configuration
```
Comment (damacus) [inline DEVELOPING.md:47]: This works for this cookbook, but if we're to extract this into a library this doesn't have great UX.

File: UPGRADING.md [hunk 7]
```diff
@@ -1,5 +1,23 @@
 # Upgrading
 
+## 10.0.0
+
+### Version 10 is a major internal rewrite
+
+- All API resource have been removed
```
Comment (damacus) [inline UPGRADING.md]:
```suggestion
- All API resources have been removed
```

File: UPGRADING.md [hunk 8]
```diff
@@ -1,5 +1,23 @@
...
+- All resource have been refactored around an abstract common base resource `_config_file`
```
Comment (damacus) [inline UPGRADING.md]:
```suggestion
- All resources have been refactored around an abstracted common base resource `_config_file`
```

File: UPGRADING.md [hunk 9]
```diff
@@ -1,5 +1,23 @@
...
+  - LDAP resources extend this via `_config_file_ldap`
+- Resource properties are automatically enumerated and added to the accumulator configuration file
+- The accumulator templates for the configuration files are now persistent via the loading of the current config state during creation
+  - Configuration is no longer automatically removed when the resource falls out of scope
+  - To remove configuration elements an explicit resource with `:delete` action is required
+    - If no properties are specified with the `:delete` action then the whole configuration section will be removed
+    - If properties are specified with the `:delete` action then only the specified properties will be removed
+  - Configuration files will not be overwritten with a blank or partial configuration when a run `raises`
```
Comment (damacus) [inline UPGRADING.md]:
```suggestion
  - Configuration files will not be overwritten with a blank or partial configuration when a run fails
```

Comment (damacus) [reply UPGRADING.md]: I think simply saying fails here is a little easier to understand

File: UPGRADING.md [hunk 10]
```diff
@@ -1,5 +1,23 @@
...
+- The auth resource has been split into separate resources
+- The log resource has been split into separate resources
+- The LDAP configuration resource have been split and renamed to singular forms
```
Comment (damacus) [inline UPGRADING.md]:
```suggestion
- The LDAP configuration resource has been split and renamed to singular forms
```

File: libraries/config.rb (hunk 2 above)
Comment (lamont-granquist) [reply libraries/config.rb:127]: this is going into the root run_context which doesn't have unified_mode

declaring compile_time resources in the root run context is really weird though, particularly to install gems.  the idempotency check here is also a codesmell.

i don't know why this isn't just `chef_gem 'deepsort'`

if it needs to be done at the outer context with idempotency around the resource and forced to compile time:

`find_resource(:chef_gem, 'deepsort') { compile_time true }`

File: DEVELOPING.md (hunk 6 above)
Comment (bmhughes) [reply DEVELOPING.md:47]: I presume you mean where you have to declare the method twice? I'm just wondering actually if the check for resource vs action class can be done in the helper and the correct method called off that.

Comment (damacus) [reply DEVELOPING.md:47]: I think what I mean here is having to add the method to the action_class for it to do something isn't super cool. 
I wonder if there's a property we can add. Or maybe I'm just thinking of this from the end user (wrapper cookbook) end which I shouldn't be. 

This property is meant to be for maintainers to add/remove. That said this pattern isn't really in use in any other part of Chef so I think it's caught me off guard slightly!

**Review by damacus — APPROVED** [review]

**Comment by kitchen-porter** [comment]
Released as: [10.0.0](https://github.com/sous-chefs/grafana/releases/tag/10.0.0)

File: DEVELOPING.md (hunk 6 above)
Comment (bmhughes) [reply DEVELOPING.md:47]: Yeh it's the latter, you'd only touch this method/hash value when developing the cookbook.

**Pull request merged**